examples/dbscan/print_timers.cpp

```diff
@@ -0,0 +1,69 @@
+/****************************************************************************
+ * Copyright (c) 2017-2021 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include <Kokkos_Core_fwd.hpp>
+
+#include <chrono>
+#include <cstdint>
+#include <iomanip>
+#include <iostream>
+#include <utility>
+#include <vector>
+
+namespace
+{
+
+struct Timer
+{
+  using clock_type = std::chrono::high_resolution_clock;
+  std::string name;
+  clock_type::time_point tick;
+  clock_type::duration duration;
+};
+
+std::vector<Timer> arborx_dbscan_example_timers;
+
+} // namespace
+
+void arborx_dbscan_example_set_create_profile_section(char const *label,
+                                                      std::uint32_t *id)
+{
+  *id = arborx_dbscan_example_timers.size();
+  arborx_dbscan_example_timers.push_back({label, {}, {}});
+}
+
+void arborx_dbscan_example_set_destroy_profile_section(std::uint32_t) {}
+
+void arborx_dbscan_example_set_start_profile_section(std::uint32_t id)
+{
+  Kokkos::fence();
+  auto now = Timer::clock_type::now();
+  auto &timer = arborx_dbscan_example_timers[id];
+  timer.tick = now;
+}
+
+void arborx_dbscan_example_set_stop_profile_section(std::uint32_t id)
+{
+  Kokkos::fence();
+  auto now = Timer::clock_type::now();
+  auto &timer = arborx_dbscan_example_timers[id];
+  timer.duration = now - timer.tick;
+}
+
+void arborx_dbscan_example_print_timers(std::ostream &os)
+{
+  for (auto const &timer : arborx_dbscan_example_timers)
+  {
+    os << "-- " << std::left << std::setw(17) << timer.name << ":      "
+       << std::chrono::duration<double>(timer.duration).count() << '\n';
+  }
```

Comment:
The only thing I don't like about this is the lack of control over the order the timers are printed. In the DBSCAN example, it makes more sense to first print construction, then query, then postprocess, and then total. However, this will be printed in the order they are created, which may be different.

I don't have a good solution. It's not a blocker, though.

Reply:
You have full control.  Current implementation prints in order section are declared but you can explicitly find labels and print them just like you did with the map.